tests/shippers/test_logstash.py

```diff
@@ -0,0 +1,193 @@
+            replay_handler.assert_not_called()
+        with self.subTest("Authentication error, request is not retried"):
+            responses.put(url=url, status=401)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            replay_handler = MagicMock(side_effect=_dummy_replay_handler)
+            logstash_shipper.set_replay_handler(replay_handler)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+            replay_handler.assert_called_once_with("logstash", {}, _dummy_event)
+
+    def test_send_https_ssl_fingerprint(self) -> None:
+        certpath = os.path.join(os.path.dirname(__file__), "ssl", "localhost.crt")
+        keypath = os.path.join(os.path.dirname(__file__), "ssl", "localhost.pkcs8.key")
+        server_address = ("localhost", 8080)
+        httpd = http.server.HTTPServer(server_address, http.server.SimpleHTTPRequestHandler)
+        httpd.socket = ssl.wrap_socket(
+            httpd.socket, server_side=True, certfile=certpath, keyfile=keypath, ssl_version=ssl.PROTOCOL_TLS
+        )
+        httpd_thread = threading.Thread(target=httpd.serve_forever)
+        httpd_thread.daemon = True
+        httpd_thread.start()
+        logstash_shipper = LogstashShipper(
+            logstash_url="https://localhost:8080",
+            ssl_assert_fingerprint="22:F7:FB:84" ":1D:43:3E" ":E7:BB:F9" ":72:F3:D8:97:AD:7C:86:E3:07:42",
+        )
+        replay_handler = MagicMock(side_effect=_dummy_replay_handler)
+        logstash_shipper.set_replay_handler(replay_handler)
+        assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+        replay_handler.assert_not_called()
+        httpd.shutdown()
+
+    @responses.activate
+    def test_send_basic_auth(self) -> None:
+        def request_callback(request: PreparedRequest) -> tuple[int, dict[Any, Any], str]:
+            _payload = []
+            usr_bytes = _username.encode("latin1")
+            pwd_bytes = _password.encode("latin1")
+            auth_string = b64encode(b":".join((usr_bytes, pwd_bytes))).decode("latin1")
+            assert request.headers["Content-Encoding"] == "gzip"
+            assert request.headers["Content-Type"] == "application/x-ndjson"
+            print(auth_string)
+            assert request.headers["Authorization"] == f"Basic {auth_string}"
+            assert request.body is not None
+            assert isinstance(request.body, bytes)
+
+            events = gzip.decompress(request.body).decode("utf-8").split("\n")
+            for event in events:
+                _payload.append(ujson.loads(event))
+
+            expected_event = _dummy_event
+            expected_event["_id"] = "_id"
+            assert _payload == [expected_event, expected_event]
+
+            return 200, {}, "okay"
+
+        def event_id_generator(event: dict[str, Any]) -> str:
+            return "_id"
+
+        url = "http://logstash_url"
+        responses.add_callback(responses.PUT, url, callback=request_callback)
```

Comment:
I'd rather use the real logstash from docker, in an integration test

Reply:
same as above